src/main/java/org/prebid/server/floors/PriceFloorFetcher.java

```diff
@@ -0,0 +1,316 @@
+        this.vertx = Objects.requireNonNull(vertx);
+        this.timeoutFactory = Objects.requireNonNull(timeoutFactory);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.mapper = Objects.requireNonNull(mapper);
+
+        fetchInProgress = new ConcurrentHashSet<>();
+        fetchedData = Caffeine.newBuilder()
+                .<String, AccountFetchContext>build()
+                .asMap();
+    }
+
+    public Future<PriceFloorRules> fetch(Account account) {
+        final PriceFloorRules cachedPriceFloorRules =
+                ObjectUtil.getIfNotNull(fetchedData.get(account.getId()), AccountFetchContext::getRules);
+
+        return cachedPriceFloorRules != null
+                ? Future.succeededFuture(cachedPriceFloorRules)
+                : fetchPriceFloorRules(account);
+    }
+
+    private Future<PriceFloorRules> fetchPriceFloorRules(Account account) {
+        if (account == null) {
+            return Future.succeededFuture();
+        }
+
+        final AccountPriceFloorsFetchConfig fetchConfig = getFetchConfig(account);
+        final Boolean fetchEnabled = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getEnabled);
+        final String fetchUrl = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getUrl);
+        final String accountId = account.getId();
+
+        if (!shouldFetchRules(fetchEnabled, fetchUrl, accountId)) {
+
+            return Future.succeededFuture();
+        }
+
+        fetchPriceFloorRulesAsynchronous(fetchConfig, accountId);
+
+        return Future.succeededFuture();
+    }
+
+    private boolean shouldFetchRules(Boolean fetchEnabled, String fetchUrl, String accountId) {
+        if (BooleanUtils.isFalse(fetchEnabled)
+                || StringUtils.isBlank(fetchUrl)
+                || fetchInProgress.contains(accountId)) {
+
+            return false;
+        }
+
+        try {
+            HttpUtil.validateUrl(fetchUrl);
+        } catch (IllegalArgumentException e) {
+            samplingLogger.warn(String.format("Malformed fetch.url passed for account %s", accountId), 0.01);
+            return false;
+        }
+
+        return true;
+    }
+
+    private static AccountPriceFloorsFetchConfig getFetchConfig(Account account) {
+        final AccountPriceFloorsConfig priceFloorsConfig =
+                ObjectUtil.getIfNotNull(account.getAuction(), AccountAuctionConfig::getPriceFloors);
+
+        return ObjectUtil.getIfNotNull(priceFloorsConfig, AccountPriceFloorsConfig::getFetch);
+    }
+
+    private void fetchPriceFloorRulesAsynchronous(AccountPriceFloorsFetchConfig fetchConfig, String accountId) {
+        final Long timeout = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getTimeout);
+        final Long maxFetchFileSizeKb =
+                ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getMaxFileSize);
+
+        fetchInProgress.add(accountId);
```

Comment:
Just to think about: should we check the progress near to downloading phase?

Reply:
Actually for what reason? We have only one blocking operation, another is java (non-blocking)code, and actually I don't see the component of code which might be interested in this information